### ISSUE TITLE
Various issues after upgrade of Crypt-CBC

### ISSUE THREAD

**Issue by wcasanova** [asbru-cm/asbru-cm#816]
**Describe the bug**
import connections fails

**To Reproduce**
1. Go to My Connections
2.  left click, select the option import connections
3.  stays permanently on, please wait, it is being imported.

**Expected behavior**
import connections, only a few are imported randomly 

**Screenshots**

**Environment (please complete the following information):**
 - OS:  archlinux
 - Ásbrú from master git version  (6.3.2.r6.g036cf2e-1)
**Additional context**

```
WARN: Adding new keybinding [Shift+Insert] for [Terminal]/[Paste selection into terminal]...
Use of uninitialized value $parent_uuid in hash element at /opt/asbru-cm/lib/PACMain.pm line 2761.
Use of uninitialized value $parent_uuid in hash element at /opt/asbru-cm/lib/PACMain.pm line 2761.
WARNING: The key derivation method "opensslv1" is deprecated. Using -pbkdf=>'pbkdf2' would be better.
Pass -nodeprecate=>1 to inhibit this message.
 at /opt/asbru-cm/lib/PACUtils.pm line 2718.
*** unhandled exception in callback:
***   Salt must be exactly 8 bytes long at /opt/asbru-cm/lib/PACUtils.pm line 2718.
***  ignoring at /usr/share/perl5/vendor_perl/Gtk3.pm line 572.
```
well it seems to be a general problem, maybe with some update of the libraries, if I want to save some changes in the preferences they are not saved.

**Comment by barkely128** [comment]
Does anybody care? New Archlinux installations aren't happy. Maybe a standalone Asbru, with all deps included can be made? We're having issues with them on a regular basis.

**Comment by barkely128** [comment]
Maybe something like this?

https://metacpan.org/pod/App::Virtualenv

**Comment by figue** [comment]
> Does anybody care? New Archlinux installations aren't happy. Maybe a standalone Asbru, with all deps included can be made? We're having issues with them on a regular basis.

What do you mean? Ásbrú works perfect in Archlinux...

**Comment by barkely128** [comment]
2 days new installation and I'm having the exact same issue the OP wrote here. Try to cleanbuild the aur package. asbru-cm.

**Comment by figue** [comment]
But this issue doesn't seem to be "a packaging problem" of Archlinux package. And the days of the installation is not relevant here.

**Comment by wcasanova** [comment]
@figue , is your archlinux up to date, have you tried doing a clean install and then importing connections, I have even tried with other computers with recent installs and it doesn't work.
maybe the least problematic thing to do would be a flatpak in this case.

**Comment by figue** [comment]
Yes I tried. Import doesn't work... But is a packaging problem? Maybe a third-party Perl library problem?

**Comment by wcasanova** [comment]
yes, it's probably a problem with Perl library updates.

**Comment by figue** [comment]
I rebuild all Perl packages in last update... And /usr/share/perl5/vendor_perl/Gtk3.pm belongs to community/perl-gtk3 
It doesn't seem to be a library update problem. Or I'm wrong?

**Comment by figue** [comment]
What Perl libraries are used to do the export/import connections?

**Comment by barkely128** [comment]
What I meant in my initial post is that it's usually something perl related. I found an AppImage package, but unfortunately it doesn't work too. Basically I'd be happy with anything, is it flakpak or snapd, doesn't matter, as long as I can have a stable and reliable Asbru all-in-one installation, cause I'm using it for my daily Linux/Unix administration tasks.

**Comment by figue** [comment]
Ok, that's we need to figure out. Maintaining perl dependencies updated is a user responsibility. One thing you can do is maintain a custom repo with all Perl packages. Can anybody that doesn't have an Arch based distro, try and see if Asbru export/import connections?

**Comment by barkely128** [comment]
Just launched full pacman upgrade, full yay (aur helper) upgrade -- no changes whatsoever. Also the problem is more serious than importing connections. When I create a new connection I immediately get this:

Use of uninitialized value $parent_uuid in hash element at /opt/asbru/lib/PACMain.pm line 2761.
WARNING: The key derivation method "opensslv1" is deprecated. Using -pbkdf=>'pbkdf2' would be better.
Pass -nodeprecate=>1 to inhibit this message.
 at /opt/asbru/lib/PACUtils.pm line 2718.
*** unhandled exception in callback:
***   Salt must be exactly 8 bytes long at /opt/asbru/lib/PACUtils.pm line 2718.
***  ignoring at /usr/lib/perl5/5.34/vendor_perl/Glib/Object/Introspection.pm line 67.

So basically useless, as I can't save a single connection. No configuration is saved whatsoever.

Arch as rolling release is notorious for having bleeding edge packages. They can easily be newer than the ones in popular distros. So keeping the deps in supported versions isn't easily maintainable like this.

Really hope packing everything together will be something the Asbru team supports as an option.

**Comment by wcasanova** [comment]
At the moment I am using Remmina, but it lacks many features. ;-(

**Comment by figue** [comment]
The problem seems related to the crypt code, caused by perl-crypt-cbc package (in community), this was upgraded to version 3.04. Downgrading to 2.33 make Ásbrú to work fine again.

@gfrenoy when you have time, can you check this library update? Thanks

https://metacpan.org/dist/Crypt-CBC

**Comment by wcasanova** [comment]
it does indeed work by downgrading that package perl-crypt-cbc , thanks for taking the time to investigate @figue 
Flatpak has a way to handle CPAN modules.
https://github.com/flatpak/flatpak-builder-tools/tree/master/cpan

**Comment by barkely128** [comment]
Thank you! Completely fixed all issues. If anyone's reading and wondering, this is how you fix it:

sudo pacman -U https://archive.archlinux.org/packages/p/perl-crypt-cbc/perl-crypt-cbc-2.33-4-any.pkg.tar.zst

**Comment by wcasanova** [comment]
> Thank you! Completely fixed all issues. If anyone's reading and wondering, this is how you fix it:
> 
> sudo pacman -U https://archive.archlinux.org/packages/p/perl-crypt-cbc/perl-crypt-cbc-2.33-4-any.pkg.tar.zst

if you also add the package, in /etc/pacman.conf to prevent it from being updated 

`IgnorePkg   = perl-crypt-cbc`

**Comment by gfrenoy** [comment]
> @gfrenoy when you have time, can you check this library update? Thanks

Can someone try to change this line in `PACUtils.pm`:

```
my $CIPHER = Crypt::CBC->new(-key => 'PAC Manager (David Torrejon Vaquerizas, david.tv@gmail.com)', -cipher => 'Blowfish', -salt => '1', -pbkdf => 'opensslv1', -nodeprecate => 1) or die "ERROR: $!";
```

And let me know if it works with Crypt::CBC 3.04 ?

**Comment by figue** [comment]
> > @gfrenoy when you have time, can you check this library update? Thanks
> 
> Can someone try to change this line in `PACUtils.pm`:
> 
> ```
> my $CIPHER = Crypt::CBC->new(-key => 'PAC Manager (David Torrejon Vaquerizas, david.tv@gmail.com)', -cipher => 'Blowfish', -salt => '1', -pbkdf => 'opensslv1', -nodeprecate => 1) or die "ERROR: $!";
> ```
> 
> And let me know if it works with Crypt::CBC 3.04 ?

Tested. It seems to be working fine. Thanks!

**Comment by wcasanova** [comment]
thanks, working perl-crypt-cbc  3.04, last commit https://github.com/asbru-cm/asbru-cm/commit/61372995970756c913e11ff26e63990fbabb50ef

**Comment by figue** [comment]
@wcasanova you can close this issue then. Updated loki package in AUR: https://aur.archlinux.org/packages/asbru-cm-loki-git

**Comment by kjuanman** [comment]
> > @gfrenoy when you have time, can you check this library update? Thanks
> 
> Can someone try to change this line in `PACUtils.pm`:
> 
> ```
> my $CIPHER = Crypt::CBC->new(-key => 'PAC Manager (David Torrejon Vaquerizas, david.tv@gmail.com)', -cipher => 'Blowfish', -salt => '1', -pbkdf => 'opensslv1', -nodeprecate => 1) or die "ERROR: $!";
> ```
> 
> And let me know if it works with Crypt::CBC 3.04 ?

PACConfig.pm and PACMain.pm has Crypt::CBC->new too. May be should be changed too.

**Comment by gfrenoy** [comment]
Thanks all for your help, suggestions & comments !  Much appreciated :)

Let's test this a few days more on loki and I'll make the change in the master branch as well.

**Comment by wcasanova** [comment]
Hi.
the error occurs again
```Use of uninitialized value $parent_uuid in hash element at /opt/asbru-cm-loki/lib/PACMain.pm line 2743.
*** unhandled exception in callback:
***   Salt must be exactly 8 bytes long at /opt/asbru-cm-loki/lib/PACUtils.pm line 2713.
***  ignoring at /usr/share/perl5/vendor_perl/Gtk3.pm line 572.
```

**Comment by wcasanova** [comment]
to solve the problem, for the time being Downgrading to 2.33

**Comment by efferre79** [comment]
I have seen there are many commits related to the new CBC version, the salt and password for the GUI. Can the issues be considered solved? What about then making a new official release? Thanks

**Comment by gfrenoy** [comment]
Just tried the loki on a recent version of Fedora with Crypt::CBC 3.04 and all is apparently working fine.

So let's close it for new.  If someone finds anything else, please let us know.

**Comment by eslutsky** [comment]
> > @gfrenoy when you have time, can you check this library update? Thanks
> 
> Can someone try to change this line in `PACUtils.pm`:
> 
> ```
> my $CIPHER = Crypt::CBC->new(-key => 'PAC Manager (David Torrejon Vaquerizas, david.tv@gmail.com)', -cipher => 'Blowfish', -salt => '1', -pbkdf => 'opensslv1', -nodeprecate => 1) or die "ERROR: $!";
> ```
> 
> And let me know if it works with Crypt::CBC 3.04 ?
i  can confirm that this fixed the issue on  on fedora-35  (with asbru-cm-6.3.2-1.fc33.noarch)
ב
בםמכ

**Comment by gablex** [comment]
> 

This fixed my issue on Fedora 35, thank you

**Comment by vostorga** [comment]
I had to apply the "my $CIPHER" patch on 6.3.3-1 running on Ubuntu 22. 
Checking the [code ](https://github.com/asbru-cm/asbru-cm/commits/master/lib/PACUtils.pm)looks like the patch was not committed.

**Comment by gfrenoy** [comment]
@vostorga : only on the `loki` branch so far.  See #858.

**Comment by ekoaripurnomo** [comment]
> > @gfrenoy when you have time, can you check this library update? Thanks
> 
> Can someone try to change this line in `PACUtils.pm`:
> 
> ```
> my $CIPHER = Crypt::CBC->new(-key => 'PAC Manager (David Torrejon Vaquerizas, david.tv@gmail.com)', -cipher => 'Blowfish', -salt => '1', -pbkdf => 'opensslv1', -nodeprecate => 1) or die "ERROR: $!";
> ```
> 
> And let me know if it works with Crypt::CBC 3.04 ?

Thanks, this solved my issue cannot save configuration and cannot save close button to close

**Comment by fixplizz** [comment]
> Thank you! Completely fixed all issues. If anyone's reading and wondering, this is how you fix it:
> 
> sudo pacman -U https://archive.archlinux.org/packages/p/perl-crypt-cbc/perl-crypt-cbc-2.33-4-any.pkg.tar.zst

Ty. everything works!

**Comment by jojaquix** [comment]
> > @gfrenoy when you have time, can you check this library update? Thanks
> 
> Can someone try to change this line in `PACUtils.pm`:
> 
> ```
> my $CIPHER = Crypt::CBC->new(-key => 'PAC Manager (David Torrejon Vaquerizas, david.tv@gmail.com)', -cipher => 'Blowfish', -salt => '1', -pbkdf => 'opensslv1', -nodeprecate => 1) or die "ERROR: $!";
> ```
> 
> And let me know if it works with Crypt::CBC 3.04 ?

This works also for fedora 36 Ásbrú Connection Manager 6.3.3 (asbru-cm)
